core-tests/shared/src/test/scala/zio/CauseSpec.scala

```diff
@@ -169,6 +171,77 @@ object CauseSpec extends ZIOBaseSpec {
         assert(stripped)(isNone)
       }
     ),
+    suite("toString")(
+      test("not fail with StackOverflowError") {
+        @tailrec
+        def genCause[E](current: Cause[E], depth: Int): Cause[E] =
+          if (depth <= 0) {
+            current
+          } else {
+            genCause(Both(current, Empty), depth - 1)
+          }
+
+        val cause = genCause(Cause.fail("Error"), 2000)
```

Comment:
This test passes with the old implementation on my machine, let's increase the depth just to make sure we're testing things correctly

```suggestion
        val cause = genCause(Cause.fail("Error"), 20000)
```